File: ios/ExpoDocumentScannerModule.swift
import ExpoModulesCore
import VisionKit
import Vision

public class ExpoDocumentScannerModule: Module, DocumentScannerModuleDelegate {
    private var promise: Promise?
    
    public func onCancel(_ result: [String: Any]) {
        promise?.resolve(result)
        promise = nil
    }
    
    public func onError(_ error: any Error) {
        promise?.reject(error)
        promise = nil
    }
    
    public func onSuccess(_ result: [String: Any]) {
        promise?.resolve(result)
        promise = nil
    }
    
    let documentScannerImpl = DocumentScannerImpl()
    let documentScannerPermissionsImpl = DocumentScannerPermissionsImpl()
    
    public func definition() -> ModuleDefinition {
        Name("ExpoDocumentScanner")
        
        OnCreate {
            documentScannerImpl.delegate = self
        }
        
        Function("checkCameraPermissions") {
            return documentScannerPermissionsImpl.checkCameraPermission()
        }
        
        AsyncFunction("requestCameraPermissions") { (promise: Promise) in
            documentScannerPermissionsImpl.checkAndRequestCameraPermission { granted in
                promise.resolve(granted)
            }
        }
        
        AsyncFunction("scanDocument") { (promise: Promise) in
            self.promise = promise
            documentScannerImpl.scanDocument()
        }.runOnQueue(.main)
    }
}

File: android/src/main/java/expo/modules/documentscanner/ExpoDocumentScannerModule.kt
package expo.modules.documentscanner

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoDocumentScannerModule : Module() {

  override fun definition() = ModuleDefinition {

    val documentScannerImpl = DocumentScannerImpl(appContext)


    Name("ExpoDocumentScanner")

    OnDestroy {
      documentScannerImpl.unregisterDocumentScanner()
    }

      AsyncFunction("scanDocument") { promise: Promise ->
        documentScannerImpl.scanDocuments(promise)
    }
  }
}
